src/preact/component/3p-frame.js

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '..';
+import {ContainWrapper} from './contain';
+import {useEffect, useLayoutEffect, useRef, useState} from '..';
+
+/** @type {!Object<string,number>} Number of 3p frames for that type. */
+export const count = {};
+
+/** @enum {string} */
+const MessageType = {
+  // TODO(wg-bento): Add more types as they become needed.
+  EMBED_SIZE: 'embed-size',
+};
+
+// Block synchronous XHR in ad. These are very rare, but super bad for UX
+// as they block the UI thread for the arbitrary amount of time until the
+// request completes.
+const BLOCK_SYNC_XHR = "sync-xhr 'none';";
+
+/**
+ * @param {*} message
+ * @return {*}
+ */
+const DEFAULT_DESERIALIZE_MESSAGE = (message) => message;
+/**
+ * Creates the iframe for the embed. Applies correct size and passes the embed
+ * attributes to the frame via JSON inside the fragment.
+ * @param {!IframeProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function IframeEmbed({
```

Comment:
I've renamed it `ProxyIframeEmbed` and created `IframeEmbed` by refactoring amp instagram. Made this PR pretty huge but though it would be valuable to keep these changes altogether for a birdseye view of how they are working together. Happy to create a separate PR with just the `IframeEmbed`/`Instagram` refactor once things look okay.